### PR TITLE
SNOW-900244-Expose DataFrame._session and Session._session_id to users

### PR DESCRIPTION
**Pull request by sfc-gh-yuwang**

### CODE REVIEW

**Comment by sfc-gh-yuwang** [comment]
should we let customer have access to session.connection? Because ServerConnection is a internal class and have no documentation on it, customer would see nothing with the hyperlink in docments

File: src/snowflake/snowpark/session.py [hunk 1]
```diff
@@ -1681,6 +1681,18 @@ def read(self) -> "DataFrameReader":
         supported sources (e.g. a file in a stage) as a DataFrame."""
         return DataFrameReader(self)
 
+    @property
+    def session_id(self) -> int:
+        """Returns an integer that represent the session id of this session."""
+        return self._session_id
+
+    @property
+    def connection(self) -> "ServerConnection":
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/session.py:1690]: We want to expose `self._conn._conn`, which is a `SnowflakeConnection` of Python connector. The ServerConnection is not a public API.

File: src/snowflake/snowpark/dataframe.py [hunk 2]
```diff
@@ -3320,6 +3320,13 @@ def na(self) -> DataFrameNaFunctions:
         """
         return self._na
 
+    @property
+    def session(self) -> "snowflake.snowpark.Session":
+        """
+        Returns a :class:`snowflake.snowpark.Session` object that provides access to the session current dataframe is relying on.
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/dataframe.py:3326]: Nit: "the current DataFrame"

File: src/snowflake/snowpark/session.py [hunk 3]
```diff
@@ -1681,6 +1681,18 @@ def read(self) -> "DataFrameReader":
         supported sources (e.g. a file in a stage) as a DataFrame."""
         return DataFrameReader(self)
 
+    @property
+    def session_id(self) -> int:
+        """Returns an integer that represent the session id of this session."""
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/session.py:1686]: Nit: "represents"

File: src/snowflake/snowpark/session.py [hunk 4]
```diff
@@ -1681,6 +1681,18 @@ def read(self) -> "DataFrameReader":
...
+        return self._session_id
+
+    @property
+    def connection(self) -> "ServerConnection":
+        """Returns a :class:`ServerConnection` object that allow you to access the connection between current session
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/session.py:1691]: "Returns the underlying :class:`SnowflakeConnection` object.

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,10 @@
 ### New Features
 
 - Accept `Iterable` objects input for `schema` when creating dataframes using `Session.create_dataframe`.
+- Added the property `DataFrame.session` to return a session object.
+- Added the property `Session.session_id` to return a integer represent session id.
```
Comment (sfc-gh-yixie) [inline CHANGELOG.md:9]: "an integer that represents the session id"

**Review by sfc-gh-yixie — CHANGES_REQUESTED** [review]

**Comment by sfc-gh-yixie** [comment]
Better add some mock unit tests.

File: src/snowflake/snowpark/session.py (hunk 1 above)
Comment (sfc-gh-yuwang) [reply src/snowflake/snowpark/session.py:1690]: Got it, I thought it was self._conn that I should expose

**Comment by sfc-gh-yuwang** [comment]
> Better add some mock unit tests.

should I add test under tests/unit/test_session.py and tests/unit/test_dataframe.py? I didn't see other test for properties in the repo

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,10 @@
 - Added support for VOLATILE/IMMUTABLE keyword when registering UDFs.
 - Added support for specifying clustering keys when saving dataframes using `DataFrame.save_as_table`.
 - Accept `Iterable` objects input for `schema` when creating dataframes using `Session.create_dataframe`.
+- Added the property `DataFrame.session` to return a session object.
+- Added the property `Session.session_id` to return an integer that represents session id.
+- Added the property `Session.connection` to return a ServerConnection object .
```
Comment (sfc-gh-sfan) [inline CHANGELOG.md:12]: I don't think we need "the property" given it is implied in the method name. What do you think?

**Review by sfc-gh-sfan — COMMENTED** [review]
Please make the PR title more descriptive.

File: CHANGELOG.md (hunk 6 above)
Comment (sfc-gh-yuwang) [reply CHANGELOG.md:12]: I think we could keep it here, it is more friendly to those who are not familiar with Python.

Comment (sfc-gh-yixie) [reply CHANGELOG.md]: This isn't accurate. The returned connection is a `SnowflakeConnection`.
The `ServerConnection` isn't exposed.

File: tests/unit/test_dataframe.py [hunk 7]
```diff
@@ -292,3 +293,14 @@ def test_dataFrame_printSchema(capfd):
         out
         == "root\n |-- A: IntegerType() (nullable = False)\n |-- B: StringType() (nullable = True)\n"
     )
+
+
+def test_session():
```
Comment (sfc-gh-yixie) [inline tests/unit/test_dataframe.py:298]: Will you add the mock for `session.connection`?

File: CHANGELOG.md [hunk 8]
```diff
@@ -7,6 +7,10 @@
 - Added support for VOLATILE/IMMUTABLE keyword when registering UDFs.
 - Added support for specifying clustering keys when saving dataframes using `DataFrame.save_as_table`.
 - Accept `Iterable` objects input for `schema` when creating dataframes using `Session.create_dataframe`.
+- Added the property `DataFrame.session` to return a `Session` object.
+- Added the property `Session.session_id` to return an integer that represents session id.
```
Comment (sfc-gh-mabrennan) [inline CHANGELOG.md:11]: id -> ID

File: src/snowflake/snowpark/session.py [hunk 9]
```diff
@@ -1695,6 +1695,18 @@ def read(self) -> "DataFrameReader":
         supported sources (e.g. a file in a stage) as a DataFrame."""
         return DataFrameReader(self)
 
+    @property
+    def session_id(self) -> int:
+        """Returns an integer that represents the session id of this session."""
```
Comment (sfc-gh-mabrennan) [inline src/snowflake/snowpark/session.py:1700]: id -> ID

File: src/snowflake/snowpark/session.py [hunk 10]
```diff
@@ -1695,6 +1695,18 @@ def read(self) -> "DataFrameReader":
...
+        return self._session_id
+
+    @property
+    def connection(self) -> "SnowflakeConnection":
+        """Returns a :class:`SnowflakeConnection` object that allow you to access the connection between current session
```
Comment (sfc-gh-mabrennan) [inline src/snowflake/snowpark/session.py:1705]: allow -> allows
between current session -> between the current session

Add a period at the end of the sentence.

File: CHANGELOG.md (hunk 6 above)
Comment (sfc-gh-yuwang) [reply CHANGELOG.md]: will modify it

File: tests/unit/test_dataframe.py (hunk 7 above)
Comment (sfc-gh-yuwang) [reply tests/unit/test_dataframe.py:298]: Yes, I just figure out how to make a mock unit test, will add all of them today

**Pull request merged**